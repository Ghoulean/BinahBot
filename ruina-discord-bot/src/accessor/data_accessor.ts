import {
    DecoratedAbnoPage,
    Localization,
    LookupResult,
} from "@ghoulean/ruina-common";
import * as fastestLevenshtein from "fastest-levenshtein";
import * as __AUTOCOMPLETE from "../data/autocomplete.json";
import * as __CN_ABNO from "../data/cn/abno.json";
import * as __EN_ABNO from "../data/en/abno.json";
import * as __JP_ABNO from "../data/jp/abno.json";
import * as __KR_ABNO from "../data/kr/abno.json";
import * as __LOOKUP_RESULTS from "../data/queryLookupResults.json";
import * as __TRCN_ABNO from "../data/trcn/abno.json";

type QueryToLookupResult = {
    [key: string]: LookupResult[];
};

type QueryToDecoratedAbnopage = {
    [key: string]: DecoratedAbnoPage;
};

type LocalizationToQueryDecoratedAbnopage = {
    [key in Localization]: QueryToDecoratedAbnopage;
};

type LevenshteinResults = {
    best: string;
    score: number;
};

const LOOKUP_RESULTS: QueryToLookupResult =
    __LOOKUP_RESULTS as QueryToLookupResult;

const CN_ABNO: QueryToDecoratedAbnopage = __CN_ABNO as QueryToDecoratedAbnopage;
const EN_ABNO: QueryToDecoratedAbnopage = __EN_ABNO as QueryToDecoratedAbnopage;
const JP_ABNO: QueryToDecoratedAbnopage = __JP_ABNO as QueryToDecoratedAbnopage;
const KR_ABNO: QueryToDecoratedAbnopage = __KR_ABNO as QueryToDecoratedAbnopage;
const TRCN_ABNO: QueryToDecoratedAbnopage =
    __TRCN_ABNO as QueryToDecoratedAbnopage;

const LOCALIZATION_TO_DECORATED_ABNO_PAGE: LocalizationToQueryDecoratedAbnopage =
    {
        [Localization.CHINESE_SIMPLIFIED]: CN_ABNO,
        [Localization.CHINESE_TRADITIONAL]: TRCN_ABNO,
        [Localization.ENGLISH]: EN_ABNO,
        [Localization.JAPANESE]: JP_ABNO,
        [Localization.KOREAN]: KR_ABNO,
    };

const AUTOCOMPLETE: string[] = __AUTOCOMPLETE.data;

const FUZZY_MATCHING_DISTANCE = 2;

export class DataAccessor {
    constructor() {}

    public lookup(query: string, preferredLocale: Localization): LookupResult {
        let lookupResults: LookupResult[] | undefined = LOOKUP_RESULTS[query];
        if (!lookupResults) {
            const levenshteinResults: LevenshteinResults =
                this.levenshteinLookup(query);
            if (levenshteinResults.score > FUZZY_MATCHING_DISTANCE) {
                throw new Error(`Couldn't identify query result for ${query}.`);
            }
            lookupResults = LOOKUP_RESULTS[levenshteinResults.best];
        }
        for (const lookupResult of lookupResults) {
            if (lookupResult.locale == preferredLocale) {
                return lookupResult;
            }
        }
        return lookupResults[0];
    }

    public getDecoratedAbnoPage(
        pageId: string,
        locale: Localization
    ): DecoratedAbnoPage {
        const decoratedAbnoPage: DecoratedAbnoPage | undefined =
            this.getLocaledAbnoMapping(locale)[pageId];
        if (!decoratedAbnoPage) {
            throw new Error(`${pageId} not found in ${locale} locale`);
        }
        return decoratedAbnoPage;
    }

    public autocomplete(query: string): string[] {
        const retVal: string[] = [];
        const cleanQuery = this.cleanQuery(query);
        for (const q of AUTOCOMPLETE) {
            if (
                this.autocompleteDistance(cleanQuery, q) <=
                FUZZY_MATCHING_DISTANCE
            ) {
                retVal.push(q);
            }
        }
        return retVal.sort((a, b) => {
            return (
                this.autocompleteDistance(cleanQuery, a) -
                    this.autocompleteDistance(cleanQuery, b) ||
                a.length - b.length ||
                a.localeCompare(b)
            );
        });
    }

    private getLocaledAbnoMapping(
        locale: Localization
    ): QueryToDecoratedAbnopage {
        return LOCALIZATION_TO_DECORATED_ABNO_PAGE[locale];
    }

    private cleanQuery(s: string): string {
        return s.replace(/\s/g, "").toLowerCase();
    }

    private autocompleteDistance(query: string, lookup: string): number {
        return fastestLevenshtein.distance(
            query,
            this.cleanQuery(lookup).substring(0, query.length)
        );
    }

    private levenshteinLookup(s: string): LevenshteinResults {
        const closest: string = fastestLevenshtein.closest(s, AUTOCOMPLETE);
        return {
            best: closest,
            score: fastestLevenshtein.distance(
                this.cleanQuery(s),
                this.cleanQuery(closest)
            ),
        };
    }
}
