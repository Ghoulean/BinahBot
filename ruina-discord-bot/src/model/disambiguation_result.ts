import { Localization, LookupResult } from "@ghoulean/ruina-common";

// TODO: This model is duplicated between the reparser and this package.
// Reconsider where it should be placed
export type DisambiguationResults = {
    id: string;
    locale: Localization;
    query: string;
    lookupResults: LookupResult[];
};