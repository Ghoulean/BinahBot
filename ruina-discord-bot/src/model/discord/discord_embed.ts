/**
 * Object holding embed data.
 * 
 * See also: https://discord.com/developers/docs/resources/channel#embed-object
 */
export type DiscordEmbed = {
    title?: string;
    description?: string;
    color?: number;
    image?: {
        url: string;
    },
    footer?: {
        text: string;
        icon_url?: string;
    },
    author?: {
        name: string;
        url?: string;
        icon_url?: string;
    },
    fields: DiscordEmbedFields[];
}

export type DiscordEmbedFields = {
    name: string;
    value: string;
    inline?: boolean;
}
