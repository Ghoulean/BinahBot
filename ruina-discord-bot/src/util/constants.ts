export enum DiscordInteractionTypes {
    PING = 1,
    APPLICATION_COMMAND = 2,
    MESSAGE_COMPONENT = 3,
    APPLICATION_COMMAND_AUTOCOMPLETE = 4,
    MODAL_SUBMIT = 5
}

export enum DiscordEmbedColors {
    AWAKENING_ABNO_PAGE = 0x40ce78,
    BREAKDOWN_ABNO_PAGE = 0xd14141
}
