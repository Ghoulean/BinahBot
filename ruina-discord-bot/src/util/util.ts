import { Localization } from "@ghoulean/ruina-common";

const DISCORD_LOCALE_TO_LOCALIZATION_MAP: { [key: string]: Localization } = {
    ko: Localization.KOREAN,
    ja: Localization.JAPANESE,
    "zh-CN": Localization.CHINESE_SIMPLIFIED,
    "zh-TW": Localization.CHINESE_TRADITIONAL,
    "en-US": Localization.ENGLISH,
    "en-GB": Localization.ENGLISH,
};

export abstract class Util {
    public static deserializeDiscordLocale(
        discordLocale: string
    ): Localization {
        return (
            DISCORD_LOCALE_TO_LOCALIZATION_MAP[discordLocale] ??
            Localization.ENGLISH
        );
    }
}
