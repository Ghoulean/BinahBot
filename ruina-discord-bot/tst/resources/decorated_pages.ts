import {
    AbnoPageSelectType,
    Chapter,
    DecoratedAbnoPage,
    DecoratedCombatPage,
    DecoratedKeyPage,
    DecoratedPassive,
    DieType,
    Ego,
    Floor,
    Localization,
    Range,
    Rarity,
    Resistance,
} from "@ghoulean/ruina-common";

export const DECORATED_ABNO_PAGE: DecoratedAbnoPage = {
    locale: Localization.ENGLISH,
    name: "TestAbnoPage",
    description: "TestDescription",
    flavorText: "TestFlavorText",
    imagePath: "test.png",
    nameId: "test",
    floor: Floor.NONE,
    chapter: Chapter.LIBRARY_OF_RUINA,
    emoLevel: 1,
    emotionRate: 1,
    emotionSign: 1,
    targetType: AbnoPageSelectType.SELECT_ONE,
    scriptId: "testScriptId",
};

export const BASE_DECORATED_COMBAT_PAGE: DecoratedCombatPage = {
    locale: Localization.ENGLISH,
    name: "name",
    description: "description",
    diceDescriptions: ["description1"],
    id: "id",
    ego: Ego.NOT_EGO,
    egoFloor: Floor.BINAH,
    scriptId: "scriptId",
    rarity: Rarity.PAPERBACK,
    dice: [
        {
            type: DieType.BLUNT,
            min: 1,
            max: 2,
            scriptId: "scriptId",
        },
    ],
    chapter: Chapter.LIBRARY_OF_RUINA,
    cost: 0,
    range: Range.MELEE,
    imagePath: "image.png",
};

export const BASE_DECORATED_KEY_PAGE: DecoratedKeyPage = {
    name: "Keypage Name",
    passiveNames: ["passiveName1"],
    locale: Localization.ENGLISH,
    id: "pageId",
    textId: "textId",
    hp: 1,
    stagger: 2,
    hpResistances: {
        [DieType.SLASH]: Resistance.FATAL,
        [DieType.PIERCE]: Resistance.WEAK,
        [DieType.BLUNT]: Resistance.NORMAL
    },
    staggerResistances: {
        [DieType.SLASH]: Resistance.ENDURED,
        [DieType.PIERCE]: Resistance.INEFFECTIVE,
        [DieType.BLUNT]: Resistance.IMMUNE
    },
    minBaseSpeed: 1,
    maxBaseSpeed: 2,
    baseLight: 3,
    chapter: Chapter.LIBRARY_OF_RUINA,
    rarity: Rarity.PAPERBACK,
    defaultPassiveIds: ["passiveId1", "passiveId2"],
    exclusiveCardsIds: ["cardId1"]
}

export const BASE_DECORATED_PASSIVE: DecoratedPassive = {
    name: "Passive Name",
    description: "Description of passive",
    locale: Localization.ENGLISH,
    id: "passiveId",
    cost: 2,
    rarity: Rarity.PAPERBACK,
    canGivePassive: false
}