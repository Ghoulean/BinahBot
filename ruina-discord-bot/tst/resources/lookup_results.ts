import {
    Chapter,
    Localization,
    LookupResult,
    PageType,
} from "@ghoulean/ruina-common";
import { DisambiguationResults } from "../../src/model/disambiguation_result";

export const ABNO_LOOKUP_RESULT: LookupResult = {
    query: "query",
    chapter: Chapter.LIBRARY_OF_RUINA,
    locale: Localization.ENGLISH,
    pageType: PageType.ABNO_PAGE,
    pageId: "pageId",
};

export const COMBAT_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.COMBAT_PAGE,
};

export const KEYPAGE_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.KEY_PAGE,
};

export const PASSIVE_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.PASSIVE,
};

export const DISAMBIGUATION_LOOKUP_RESULT: LookupResult = {
    ...ABNO_LOOKUP_RESULT,
    pageType: PageType.DISAMBIGUATION,
};

export const DISAMBIGUATION_RESULTS: DisambiguationResults = {
    id: "disambiguation_id",
    locale: Localization.ENGLISH,
    query: "disambiguation",
    lookupResults: [ABNO_LOOKUP_RESULT, COMBAT_LOOKUP_RESULT],
};
