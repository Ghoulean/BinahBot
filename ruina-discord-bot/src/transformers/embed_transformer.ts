import {
    DecoratedAbnoPage,
    DecoratedCombatPage,
    DecoratedKeyPage,
    DecoratedPassive,
    Die,
    Localization,
    LookupResult,
    Rarity,
    dieTypeToShortString,
} from "@ghoulean/ruina-common";
import { DisambiguationResults } from "../model/disambiguation_result";
import {
    DiscordEmbed,
    DiscordEmbedFields,
} from "../model/discord/discord_embed";
import { DiscordEmbedColors } from "../util/constants";

export class EmbedTransformer {
    private readonly s3BucketName: string;

    constructor(s3BucketName: string) {
        this.s3BucketName = s3BucketName;
    }

    // TODO: localize via _requestLocale
    public transformAbnoPage(
        abno: DecoratedAbnoPage,
        _requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number =
            abno.emotionSign > 0
                ? DiscordEmbedColors.AWAKENING_ABNO_PAGE
                : DiscordEmbedColors.BREAKDOWN_ABNO_PAGE;
        const abnoType: string =
            abno.emotionSign > 0 ? "Awakening" : "Breakdown";
        return {
            title: abno.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${abno.imagePath}`,
            },
            fields: [
                {
                    name: "Flavor text",
                    value: abno.flavorText,
                    inline: true,
                },
                {
                    name: "Effect",
                    value: abno.description,
                    inline: true,
                },
                {
                    name: "Bias",
                    value: String(abno.emotionRate),
                    inline: true,
                },
                {
                    name: "Type",
                    value: abnoType,
                    inline: true,
                },
                {
                    name: "Emotion level",
                    value: String(abno.emoLevel),
                    inline: true,
                },
                {
                    name: "Floor",
                    value: abno.floor,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformCombatPage(
        combat: DecoratedCombatPage,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(combat.rarity);
        const fields: DiscordEmbedFields[] = [
            {
                name: "Cost",
                value: String(combat.cost),
                inline: true,
            },
            {
                name: "Range",
                value: combat.range,
                inline: true,
            },
            {
                name: "Rarity",
                value: combat.rarity,
                inline: true,
            },
        ];
        if (combat.description) {
            fields.push({
                name: "Page Description",
                value: combat.description,
                inline: true,
            });
        }

        fields.push({
            name: "Dice",
            value: this.mapDie(
                combat.dice,
                combat.diceDescriptions,
                requestLocale
            ),
            inline: false,
        });
        return {
            title: combat.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${combat.imagePath}`,
            },
            fields: fields,
        };
    }
    // TODO: localize via _requestLocale
    public transformKeyPage(
        keyPage: DecoratedKeyPage,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(keyPage.rarity);

        return {
            title: keyPage.name,
            color: embedColor,
            fields: [
                {
                    name: "Rarity",
                    value: keyPage.rarity,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformPassive(
        passive: DecoratedPassive,
        requestLocale: Localization
    ): DiscordEmbed {
        const embedColor: number = this.rarityToColor(passive.rarity);

        return {
            title: passive.name,
            color: embedColor,
            fields: [
                {
                    name: "Cost",
                    value: String(passive.cost),
                    inline: true,
                },
                {
                    name: "Rarity",
                    value: passive.rarity,
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public transformDisambiguationPage(
        disambiguation: DisambiguationResults,
        _requestLocale: Localization
    ): DiscordEmbed {
        return {
            fields: [
                {
                    name: `**${disambiguation.query}** may refer to:`,
                    value: this.mapToIndentedList(
                        disambiguation.lookupResults.map(
                            (lookupResult: LookupResult) => {
                                return lookupResult.query;
                            }
                        )
                    ),
                    inline: true,
                },
            ],
        };
    }

    // TODO: localize via _requestLocale
    public noResultsFoundEmbed(
        query: string,
        _requestLocale: Localization
    ): DiscordEmbed {
        return {
            description: `*No results found for \`${query}\`.*`,
            fields: [],
        };
    }

    private rarityToColor(rarity: Rarity): number {
        switch (rarity) {
            case Rarity.PAPERBACK:
                return DiscordEmbedColors.PAPERBACK_RARITY;
            case Rarity.HARDCOVER:
                return DiscordEmbedColors.HARDCOVER_RARITY;
            case Rarity.LIMITED:
                return DiscordEmbedColors.LIMITED_RARITY;
            case Rarity.OBJET_D_ART:
                return DiscordEmbedColors.OBJET_D_ART_RARITY;
        }
    }

    private mapDie(
        dice: Die[],
        diceDescriptions: string[],
        locale: Localization
    ): string {
        return this.mapToIndentedList(
            dice.map((die: Die, index: number) => {
                return `${dieTypeToShortString(die.type, locale)} ${die.min}-${
                    die.max
                } ${diceDescriptions[index]}`;
            })
        );
    }

    private mapToIndentedList(arr: string[]): string {
        return arr
            .map((str: string) => {
                return ` > - ${str}`;
            })
            .join("\n");
    }
}
