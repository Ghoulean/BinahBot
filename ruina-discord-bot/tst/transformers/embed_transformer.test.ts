import {
    Chapter,
    DecoratedAbnoPage,
    Localization,
    PageType,
    Rarity,
} from "@ghoulean/ruina-common";
import { DataAccessor } from "../../src/accessor/data_accessor";
import { DisambiguationResults } from "../../src/model/disambiguation_result";
import { DiscordEmbed } from "../../src/model/discord/discord_embed";
import { EmbedTransformer } from "../../src/transformers/embed_transformer";
import { DiscordEmbedColors } from "../../src/util/constants";
import {
    BASE_DECORATED_COMBAT_PAGE,
    DECORATED_ABNO_PAGE,
} from "../resources/decorated_pages";
import { DISAMBIGUATION_RESULTS } from "../resources/lookup_results";

const DEFAULT_REQUEST_LOCALE: Localization = Localization.ENGLISH;

const POSITIVE_DECORATED_ABNO_PAGE: DecoratedAbnoPage = {
    ...DECORATED_ABNO_PAGE,
    emotionRate: 1,
    emotionSign: 1,
};

const NEGATIVE_DECORATED_ABNO_PAGE: DecoratedAbnoPage = {
    ...DECORATED_ABNO_PAGE,
    emotionSign: -1,
    emotionRate: -1,
};

const NO_RESULTS_FOUND_QUERY: string = "query";

const S3_BUCKET_NAME: string = "s3bucketname";

let embedTransformer: EmbedTransformer;

beforeEach(() => {
    embedTransformer = new EmbedTransformer(S3_BUCKET_NAME);
});

test("should transform postive decorated abno page to Discord embed", () => {
    const discordEmbed: DiscordEmbed = embedTransformer.transformAbnoPage(
        POSITIVE_DECORATED_ABNO_PAGE,
        DEFAULT_REQUEST_LOCALE
    );
    expect(discordEmbed.color).toBe(DiscordEmbedColors.AWAKENING_ABNO_PAGE);
    expect(discordEmbed).toMatchSnapshot();
});

test("should transform negative decorated abno page to Discord embed", () => {
    const discordEmbed: DiscordEmbed = embedTransformer.transformAbnoPage(
        NEGATIVE_DECORATED_ABNO_PAGE,
        DEFAULT_REQUEST_LOCALE
    );
    expect(discordEmbed.color).toBe(DiscordEmbedColors.BREAKDOWN_ABNO_PAGE);
    expect(discordEmbed).toMatchSnapshot();
});

test.each([
    {
        descriptor: "rarity",
        combatPage: {
            ...BASE_DECORATED_COMBAT_PAGE,
            rarity: Rarity.PAPERBACK,
        },
        expectedColor: DiscordEmbedColors.PAPERBACK_RARITY,
    },
    {
        descriptor: "rarity",
        combatPage: {
            ...BASE_DECORATED_COMBAT_PAGE,
            rarity: Rarity.HARDCOVER,
        },
        expectedColor: DiscordEmbedColors.HARDCOVER_RARITY,
    },
    {
        descriptor: "rarity",
        combatPage: {
            ...BASE_DECORATED_COMBAT_PAGE,
            rarity: Rarity.LIMITED,
        },
        expectedColor: DiscordEmbedColors.LIMITED_RARITY,
    },
    {
        descriptor: "rarity",
        combatPage: {
            ...BASE_DECORATED_COMBAT_PAGE,
            rarity: Rarity.OBJET_D_ART,
        },
        expectedColor: DiscordEmbedColors.OBJET_D_ART_RARITY,
    },
])(
    "should transform decorated combat page to Discord embed (rarity: $rarity)",
    ({ combatPage, expectedColor }) => {
        const discordEmbed: DiscordEmbed = embedTransformer.transformCombatPage(
            combatPage,
            DEFAULT_REQUEST_LOCALE
        );
        expect(discordEmbed.color).toBe(expectedColor);
        expect(discordEmbed).toMatchSnapshot();
    }
);

test("should transform disambiguation page to Discord embed", () => {
    const discordEmbed: DiscordEmbed = embedTransformer.transformDisambiguationPage(
        DISAMBIGUATION_RESULTS,
        DEFAULT_REQUEST_LOCALE
    );
    expect(discordEmbed).toMatchSnapshot();
});

test("should return no results found embed", () => {
    expect(
        embedTransformer.noResultsFoundEmbed(
            NO_RESULTS_FOUND_QUERY,
            DEFAULT_REQUEST_LOCALE
        )
    ).toMatchSnapshot();
});
