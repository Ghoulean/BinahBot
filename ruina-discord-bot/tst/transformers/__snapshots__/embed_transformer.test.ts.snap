// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return no results found embed 1`] = `
{
  "description": "*No results found for \`query\`.*",
  "fields": [],
}
`;

exports[`should transform decorated combat page to Discord embed (rarity: $rarity) 1`] = `
{
  "color": 8115825,
  "fields": [
    {
      "inline": true,
      "name": "Cost",
      "value": "0",
    },
    {
      "inline": true,
      "name": "Range",
      "value": "Near",
    },
    {
      "inline": true,
      "name": "Rarity",
      "value": "Common",
    },
    {
      "inline": true,
      "name": "Description",
      "value": "description",
    },
    {
      "inline": true,
      "name": "Dice",
      "value": "description",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comimage.png",
  },
  "title": "name",
}
`;

exports[`should transform decorated combat page to Discord embed (rarity: $rarity) 2`] = `
{
  "color": 3170234,
  "fields": [
    {
      "inline": true,
      "name": "Cost",
      "value": "0",
    },
    {
      "inline": true,
      "name": "Range",
      "value": "Near",
    },
    {
      "inline": true,
      "name": "Rarity",
      "value": "Uncommon",
    },
    {
      "inline": true,
      "name": "Description",
      "value": "description",
    },
    {
      "inline": true,
      "name": "Dice",
      "value": "description",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comimage.png",
  },
  "title": "name",
}
`;

exports[`should transform decorated combat page to Discord embed (rarity: $rarity) 3`] = `
{
  "color": 7022271,
  "fields": [
    {
      "inline": true,
      "name": "Cost",
      "value": "0",
    },
    {
      "inline": true,
      "name": "Range",
      "value": "Near",
    },
    {
      "inline": true,
      "name": "Rarity",
      "value": "Rare",
    },
    {
      "inline": true,
      "name": "Description",
      "value": "description",
    },
    {
      "inline": true,
      "name": "Dice",
      "value": "description",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comimage.png",
  },
  "title": "name",
}
`;

exports[`should transform decorated combat page to Discord embed (rarity: $rarity) 4`] = `
{
  "color": 15449600,
  "fields": [
    {
      "inline": true,
      "name": "Cost",
      "value": "0",
    },
    {
      "inline": true,
      "name": "Range",
      "value": "Near",
    },
    {
      "inline": true,
      "name": "Rarity",
      "value": "Unique",
    },
    {
      "inline": true,
      "name": "Description",
      "value": "description",
    },
    {
      "inline": true,
      "name": "Dice",
      "value": "description",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comimage.png",
  },
  "title": "name",
}
`;

exports[`should transform disambiguation page to Discord embed 1`] = `
{
  "fields": [
    {
      "inline": true,
      "name": "**disambiguation** may refer to:",
      "value": " > - query
 > - query",
    },
  ],
}
`;

exports[`should transform negative decorated abno page to Discord embed 1`] = `
{
  "color": 13713729,
  "fields": [
    {
      "inline": true,
      "name": "Flavor text",
      "value": "TestFlavorText",
    },
    {
      "inline": true,
      "name": "Effect",
      "value": "TestDescription",
    },
    {
      "inline": true,
      "name": "Bias",
      "value": "-1",
    },
    {
      "inline": true,
      "name": "Type",
      "value": "Breakdown",
    },
    {
      "inline": true,
      "name": "Emotion level",
      "value": "1",
    },
    {
      "inline": true,
      "name": "Floor",
      "value": "None",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comtest.png",
  },
  "title": "TestAbnoPage",
}
`;

exports[`should transform postive decorated abno page to Discord embed 1`] = `
{
  "color": 4247160,
  "fields": [
    {
      "inline": true,
      "name": "Flavor text",
      "value": "TestFlavorText",
    },
    {
      "inline": true,
      "name": "Effect",
      "value": "TestDescription",
    },
    {
      "inline": true,
      "name": "Bias",
      "value": "1",
    },
    {
      "inline": true,
      "name": "Type",
      "value": "Awakening",
    },
    {
      "inline": true,
      "name": "Emotion level",
      "value": "1",
    },
    {
      "inline": true,
      "name": "Floor",
      "value": "None",
    },
  ],
  "image": {
    "url": "https://s3bucketname.s3.amazonaws.comtest.png",
  },
  "title": "TestAbnoPage",
}
`;
